### PR TITLE
Upgrade MaxText post-training pinned dependencies to JAX 0.11.1

### PR DESCRIPTION
**Pull request by darisoy**
# Description

Upgrades MaxText TPU post-training pinned dependencies to JAX 0.11.1 ecosystem:
- `jax==0.11.1`
- `jaxlib==0.11.1`
- `libtpu==0.0.46`
- `flax==0.12.9`
- `optax==0.2.8`
- `torch==2.11.0+cpu` and `torchvision==0.26.0+cpu`
- Added required resolution overrides (`mpmath<=1.3.0,>=1.3.0`, `llguidance<1.8.0,>=1.7.0`, `boto3>=1.34.0`, `tensorflow==2.20.0`, `tensorflow-text==2.20.1`, `tokamax>=0.0.13`, `xprof<=2.23.0`).
- Updated `generate_requirements.sh` post-training resolution flow to inject `override-dependencies` and Linux x86_64 target platform constraints.
- Updated `MaxTextCheckpointManager` in `distillation_utils.py` to correctly initialize and delegate to Tunix's `CheckpointManager` with MaxText `GrainCheckpointHandler`.

FIXES: b/552616887

cc @bvandermoon @SurbhiJainUSC

# Tests

- Unit tests in `tests/post_training/unit/distillation_checkpointing_test.py` (2 passed)
- Unit tests in `tests/post_training/unit/train_distill_test.py` (23 passed, 1 skipped)
- Unit tests in `tests/post_training/unit/dpo_trainer_correctness_test.py` (2 passed)
- Full post-training unit suite `tests/post_training/unit/` (56 passed, 271 skipped)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests and provided workload links above if applicable.
- [x] I have added or updated tests accordingly.
- [x] I have updated documentation accordingly.
- [x] I have included unit tests for new features or bug fixes.
- [x] I have added unit test and integration test coverage for all modified code paths.

### CODE REVIEW

File: src/dependencies/scripts/generate_requirements.sh [hunk 1]
```diff
@@ -99,7 +105,34 @@ strip_exact_pins() {
   sed 's/^\([A-Za-z0-9._-]*\(\[[^]]*\]\)\?\)[[:space:]]*==[[:space:]]*\([^,; ]*\)/\1>=\3/'
 }
 
+if [[ -n "$OVERRIDE_REQUIREMENTS" ]]; then
+  echo "=== Adding overrides from $OVERRIDE_REQUIREMENTS ==="
+
+  python3 -c "
```
[the remark this answers is not in the document]
Comment (darisoy) [reply src/dependencies/scripts/generate_requirements.sh]: Addressed in commit fdd06eacc using `uv run --no-project --with toml python3` to execute the Python script in an isolated ephemeral environment without host dependency requirements.

File: .github/workflows/build_package.yml [hunk 2]
```diff
@@ -43,7 +43,6 @@ jobs:
   build_and_upload:
     name: Build Wheel
     runs-on: ${{ inputs.cloud_runner != '' && inputs.cloud_runner || fromJson(format('["self-hosted", "{0}", "{1}"]', inputs.device_type, inputs.device_name)) }}
-    timeout-minutes: 15
```
Comment (bvandermoon) [inline .github/workflows/build_package.yml:46]: Why are we removing the timeouts from all these files? Were they causing an issue somewhere?

Comment (darisoy) [reply .github/workflows/build_package.yml:46]: i think this is the result of poor merge conflict management since these timeouts were recently added. fixing this now.

File: src/dependencies/scripts/generate_requirements.sh [hunk 3]
```diff
@@ -68,9 +68,15 @@ fi
...
 ARTIFACT_DIR="$OUTPUT_DIR/python${PYTHON_VERSION//./_}"
 
-echo "=== Running seed-env with $BASE_REQUIREMENTS ==="
+if [[ "$BASE_REQUIREMENTS" == *post-train* ]]; then
+  SEED_BASE_REQUIREMENTS="$(dirname "$BASE_REQUIREMENTS")/requirements.txt"
+else
+  SEED_BASE_REQUIREMENTS="$BASE_REQUIREMENTS"
+fi
+
+echo "=== Running seed-env with $SEED_BASE_REQUIREMENTS ==="
```
Comment (dipannita08) [inline src/dependencies/scripts/generate_requirements.sh]: Can we also echo when seed-env was run with post-train requirements as opposed to base requirements?

File: src/maxtext/common/grain_utility.py [hunk 4]
```diff
@@ -395,12 +405,15 @@ def restore_single_process(item, process_index, process_count):
...
         raise ValueError(f"File {filename} does not exist.")
-      state = filename.read_text()
-      if isinstance(item, grain.DatasetIterator):
-        state = json.loads(state)
-      else:
-        state = state.encode()
-      item.set_state(state)
+      state_text = filename.read_text()
+      try:
+        state = json.loads(state_text)
```
Comment (dipannita08) [inline src/maxtext/common/grain_utility.py:413]: Was this broad try shift from grain DatasetIterator to load try intended? Not familiar with the delta in the jax upgrade but might also be useful to log this for dubuggability.

**Review by dipannita08 — APPROVED** [review]

File: src/dependencies/scripts/generate_requirements.sh (hunk 3 above)
Comment (darisoy) [reply src/dependencies/scripts/generate_requirements.sh]: Added explicit logging to distinguish when `seed-env` is invoked for post-train requirements using base requirements vs. base requirements directly.

File: src/maxtext/common/grain_utility.py (hunk 4 above)
Comment (darisoy) [reply src/maxtext/common/grain_utility.py:413]: Yes, in newer Grain and PyGrain versions, iterator state formats can be either JSON dictionaries, serialized strings, or raw bytes depending on the iterator implementation and pipeline stage. Using `json.loads` with fallback to UTF-8 bytes gracefully handles all variants. Added `logging.debug(...)` calls for both JSON decoding and attribute assignment fallbacks to aid debuggability.

**Review by bvandermoon — APPROVED** [review]
